scrypto/src/math/integer.rs

```diff
@@ -0,0 +1,745 @@
+                            *self = (*self * other).try_into().unwrap();
+                        }
+                    }
+                    forward_ref_op_assign! { impl MulAssign, mul_assign for $t, $o }
+
+                    impl Div<$o> for $t {
+                        type Output = $out;
+
+                        #[inline]
+                        fn div(self, other: $o) -> $out {
+                            BigInt::from(self).div(&BigInt::from(other)).try_into().unwrap()
+                        }
+                    }
+                    forward_ref_binop! { impl Div, div for $t, $o }
+
+                    impl DivAssign<$o> for $t {
+                        #[inline]
+                        fn div_assign(&mut self, other: $o) {
+                            *self = (*self / other).try_into().unwrap();
+                        }
+                    }
+                    forward_ref_op_assign! { impl DivAssign, div_assign for $t, $o }
+
+                    impl Rem<$o> for $t {
+                        type Output = $out;
+
+                        #[inline]
+                        fn rem(self, other: $o) -> $out {
+                            BigInt::from(self).rem(&BigInt::from(other)).try_into().unwrap()
+                        }
+                    }
+                    forward_ref_binop! { impl Rem, rem for $t, $o }
+
+                    impl RemAssign<$o> for $t {
+                        #[inline]
+                        fn rem_assign(&mut self, other: $o) {
+                            *self = (*self % other).try_into().unwrap();
+                        }
+                    }
+                    forward_ref_op_assign! { impl RemAssign, rem_assign for $t, $o }
+                    impl Pow<$o> for $t {
+                        type Output = $t;
+
+                        /// Raises self to the power of `exp`, using exponentiation by squaring.
+                        ///
+                        #[inline]
+                        #[must_use = "this returns the result of the operation, \
+                              without modifying the original"]
+                        fn pow(self, other: $o) -> $t {
+                            BigInt::from(self).pow(BigUint::try_from(BigInt::from(other)).unwrap()).try_into().unwrap()
+                        }
+                    }
+
+                    forward_ref_binop! { impl Pow, pow for $t, $o }
+
+                    )*
+            }
+        };
+    }
+checked_impl! {
+//(self, other, output)
+(I8, u8, I8), (I8, u16, I16), (I8, u32, I32), (I8, u64, I64), (I8, u128, I128),
+(I8, i8, I8), (I8, i16, I16), (I8, i32, I32), (I8, i64, I64), (I8, i128, I128),
+(I8, U8, I8), (I8, U16, I16), (I8, U32, I32), (I8, U64, I64), (I8, U128, I128),
+(I8, U256, I256), (I8, U384, I384), (I8, U512, I512), (I8, I8, I8), (I8, I16, I16),
+(I8, I32, I32), (I8, I64, I64), (I8, I128, I128), (I8, I256, I256), (I8, I384, I384),
+(I8, I512, I512),
```

Comment:
While I like auto type promotion like in Java, I do have a concern about its use in a signed-unsigned-mixed context.

For example, `I8 + U256 -> I256` doesn't look right.  `0 + U256::MAX -> panic`.

Do you think we should do type promotion when the signedness matches?

Reply:
Yep, that sounds better, and I will add some shorthand notation for `.try_into().unwrap() -> .t()` and `.into() -> i()`
 to make code more readable.

Reply:
Fixed in 65d31d